[jit] Better match behavior of loaded ScriptModules vs. freshly created ones

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#43298 [jit] Better match behavior of loaded ScriptModules vs. freshly created ones**
* #43284 [jit] fix segfault in attribute lookup on loaded ScriptModules

IR emitter uses `ModuleValue` to represent ScriptModules and emit IR for
attribute access, submodule access, etc.

`ModuleValue` relies on two pieces of information, the JIT type of the
module, and the `ConcreteModuleType`, which encapsulates Python-only
information about the module.

ScriptModules loaded from a package used to create a dummy
ConcreteModuleType without any info in it. This led to divergences in
behavior during compilation.

This PR makes the two ways of constructing a ConcreteModuleType equivalent,
modulo any py-only information (which, by definition, is never present in
packaged files anyway).

Differential Revision: [D23228738](https://our.internmc.facebook.com/intern/diff/D23228738)

suo committed:
[jit] Better match behavior of loaded ScriptModules vs. freshly created ones

IR emitter uses `ModuleValue` to represent ScriptModules and emit IR for
attribute access, submodule access, etc.

`ModuleValue` relies on two pieces of information, the JIT type of the
module, and the `ConcreteModuleType`, which encapsulates Python-only
information about the module.

ScriptModules loaded from a package used to create a dummy
ConcreteModuleType without any info in it. This led to divergences in
behavior during compilation.

This PR makes the two ways of constructing a ConcreteModuleType equivalent,
modulo any py-only information (which, by definition, is never present in
packaged files anyway).

[ghstack-poisoned]

suo committed:
Update on "[jit] Better match behavior of loaded ScriptModules vs. freshly created ones"


IR emitter uses `ModuleValue` to represent ScriptModules and emit IR for
attribute access, submodule access, etc.

`ModuleValue` relies on two pieces of information, the JIT type of the
module, and the `ConcreteModuleType`, which encapsulates Python-only
information about the module.

ScriptModules loaded from a package used to create a dummy
ConcreteModuleType without any info in it. This led to divergences in
behavior during compilation.

This PR makes the two ways of constructing a ConcreteModuleType equivalent,
modulo any py-only information (which, by definition, is never present in
packaged files anyway).

[ghstack-poisoned]

suo committed:
Update on "[jit] Better match behavior of loaded ScriptModules vs. freshly created ones"


IR emitter uses `ModuleValue` to represent ScriptModules and emit IR for
attribute access, submodule access, etc.

`ModuleValue` relies on two pieces of information, the JIT type of the
module, and the `ConcreteModuleType`, which encapsulates Python-only
information about the module.

ScriptModules loaded from a package used to create a dummy
ConcreteModuleType without any info in it. This led to divergences in
behavior during compilation.

This PR makes the two ways of constructing a ConcreteModuleType equivalent,
modulo any py-only information (which, by definition, is never present in
packaged files anyway).

Differential Revision: [D23228738](https://our.internmc.facebook.com/intern/diff/D23228738)

[ghstack-poisoned]

SplitInfinity commented on torch/csrc/jit/frontend/concrete_module_type.h:
```diff
@@ -57,11 +57,14 @@ class ConcreteModuleType;
 // this to decide whether we can share types between modules.
 class VISIBILITY_HIDDEN ConcreteModuleTypeBuilder {
  public:
+  ConcreteModuleTypeBuilder() {}
```
Out of curiosity, why does this need to be public now? I don't see any new callsites for it.

Is it due to the changes to `add_constant` in `script_init.cpp`?

suo replied:
good catch, it doesn't (I moved this for a different version of this PR and didn't revert)

SplitInfinity commented on test/test_jit.py:
```diff
@@ -15182,6 +15182,38 @@ def forward(self, x):
+        class InnerSubmod(nn.Module):
+            def __init__(self):
+                super().__init__()
+                self.register_buffer("foo", torch.ones(1))
```
Consider adding a constant, parameter and regular data attribute to either `InnerSubmod` or `Inner` to make sure those code paths are tested too.

SplitInfinity commented on test/test_jit.py:
```diff
@@ -15182,6 +15182,38 @@ def forward(self, x):
+                # access inn eelem
+                return self.inner.submod(x) + self.inner.submod.foo
+
+        inner_module = torch.jit.script(Inner())
```
```suggestion
        inner_module = torch.jit.script(Inner())
        wrapped = Wrapper(inner_module)
        self.checkModule(wrapped, torch.ones(1))
```
To be more faithful to the stated goal of this PR, I think we should run `checkModule` with this version of `inner_module`. We could also compare the outputs of `Wrapped(inner_module)` and `Wrapped(inner_module_loaded)`.

suo replied:
good idea

SplitInfinity reviewed (commented):
One additional question: do we need to do anything for function attributes? I see that they are part of the `ConcreteModuleType`, but not the JIT type.

suo commented:
> One additional question: do we need to do anything for function attributes? I see that they are part of the ConcreteModuleType, but not the JIT type.

Anything that is not representable on the JIT type can't be captured here. This makes saving/loading a "lossy", since once you save/load all you have is a JIT type, and lose all the other Python-y bits like function attributes, module iterator kind, etc. The way to reduce this lossiness is by making more things representable first-class on the JIT type.

suo commented on torch/csrc/jit/frontend/concrete_module_type.cpp:
```diff
@@ -57,15 +44,43 @@ ClassTypePtr ConcreteModuleTypeBuilder::createTypeFromThis() const {
+    // Populate the builder metadata from the JIT type. This is to ensure
+    // ConcreteModuleTypes produced from Python and ones produced from a JIT
+    // type directly behave the same to the rest of the system.
+    for (size_t i = 0; i < classType->numAttributes(); i++) {
```
I think this makes the code a little less readable for no clear benefit, so I'm going to skip this one unless you feel strongly about it.

SplitInfinity replied:
I prefer my suggestion to avoid making an extra function call every time the loop exit condition is tested: https://llvm.org/docs/CodingStandards.html#don-t-evaluate-end-every-time-through-a-loop

But `numAttributes()` is not an expensive call, so I will not lose sleep over this.

suo replied:
yes, I would expect a compiler to inline and then avoid recomputing numAttributes(). I think the suggestion makes more sense if we are computing the iterator of a complicated data structure instead of returning an int. Although if the data structure is that complicated, writing a raw for loop over its iterator is probably bad to begin with…

suo committed:
Update on "[jit] Better match behavior of loaded ScriptModules vs. freshly created ones"


IR emitter uses `ModuleValue` to represent ScriptModules and emit IR for
attribute access, submodule access, etc.

`ModuleValue` relies on two pieces of information, the JIT type of the
module, and the `ConcreteModuleType`, which encapsulates Python-only
information about the module.

ScriptModules loaded from a package used to create a dummy
ConcreteModuleType without any info in it. This led to divergences in
behavior during compilation.

This PR makes the two ways of constructing a ConcreteModuleType equivalent,
modulo any py-only information (which, by definition, is never present in
packaged files anyway).

Differential Revision: [D23228738](https://our.internmc.facebook.com/intern/diff/D23228738)

[ghstack-poisoned]

SplitInfinity commented on test/test_jit.py:
```diff
@@ -15182,6 +15182,42 @@ def forward(self, x):
+                super().__init__()
+                self.register_buffer("foo", torch.ones(1))
+                self.register_parameter("bar", torch.nn.Parameter(torch.ones(1)))
+                self.baz = torch.ones(1)
```
Would this be considered a constant or a data attribute? Just want to make sure we test constants because that code path has the most changes (e.g. constants are stored as `IValue`s and not `py::object`s now).

suo replied:
forgot about constants, will add.

suo committed:
Update on "[jit] Better match behavior of loaded ScriptModules vs. freshly created ones"


IR emitter uses `ModuleValue` to represent ScriptModules and emit IR for
attribute access, submodule access, etc.

`ModuleValue` relies on two pieces of information, the JIT type of the
module, and the `ConcreteModuleType`, which encapsulates Python-only
information about the module.

ScriptModules loaded from a package used to create a dummy
ConcreteModuleType without any info in it. This led to divergences in
behavior during compilation.

This PR makes the two ways of constructing a ConcreteModuleType equivalent,
modulo any py-only information (which, by definition, is never present in
packaged files anyway).

Differential Revision: [D23228738](https://our.internmc.facebook.com/intern/diff/D23228738)

[ghstack-poisoned]

SplitInfinity reviewed (approved):
🥇

suo committed:
Update on "[jit] Better match behavior of loaded ScriptModules vs. freshly created ones"


IR emitter uses `ModuleValue` to represent ScriptModules and emit IR for
attribute access, submodule access, etc.

`ModuleValue` relies on two pieces of information, the JIT type of the
module, and the `ConcreteModuleType`, which encapsulates Python-only
information about the module.

ScriptModules loaded from a package used to create a dummy
ConcreteModuleType without any info in it. This led to divergences in
behavior during compilation.

This PR makes the two ways of constructing a ConcreteModuleType equivalent,
modulo any py-only information (which, by definition, is never present in
packaged files anyway).

Differential Revision: [D23228738](https://our.internmc.facebook.com/intern/diff/D23228738)

[ghstack-poisoned]

suo committed:
Update on "[jit] Better match behavior of loaded ScriptModules vs. freshly created ones"


IR emitter uses `ModuleValue` to represent ScriptModules and emit IR for
attribute access, submodule access, etc.

`ModuleValue` relies on two pieces of information, the JIT type of the
module, and the `ConcreteModuleType`, which encapsulates Python-only
information about the module.

ScriptModules loaded from a package used to create a dummy
ConcreteModuleType without any info in it. This led to divergences in
behavior during compilation.

This PR makes the two ways of constructing a ConcreteModuleType equivalent,
modulo any py-only information (which, by definition, is never present in
packaged files anyway).

Differential Revision: [D23228738](https://our.internmc.facebook.com/intern/diff/D23228738)

[ghstack-poisoned]

suo committed:
Update on "[jit] Better match behavior of loaded ScriptModules vs. freshly created ones"


IR emitter uses `ModuleValue` to represent ScriptModules and emit IR for
attribute access, submodule access, etc.

`ModuleValue` relies on two pieces of information, the JIT type of the
module, and the `ConcreteModuleType`, which encapsulates Python-only
information about the module.

ScriptModules loaded from a package used to create a dummy
ConcreteModuleType without any info in it. This led to divergences in
behavior during compilation.

This PR makes the two ways of constructing a ConcreteModuleType equivalent,
modulo any py-only information (which, by definition, is never present in
packaged files anyway).

Differential Revision: [D23228738](https://our.internmc.facebook.com/intern/diff/D23228738)

[ghstack-poisoned]
